### ISSUE TITLE
Unpooling layers and generative architectures

### ISSUE THREAD

**Issue by robclouth** [rusty1s/pytorch_geometric#98]
Hey there! Thanks for your hard work with this project, it's super cool. In the end of the splineCNN paper, it was mentioned that unpooling layers were a future possibility. I'd love to try generating graphs from a continuous latent space. Any ideas for where I might start? Any projects like this one that you're aware of with generative capabilities? Thanks.

**Comment by robclouth** [comment]
Oh wait I just saw the other issue on this topic. Sorry for cluttering the issues.

**Comment by robclouth** [comment]
I've done some research and this paper and code could be interesting to you as it uses decoding.

https://github.com/Ruiqi-Hu/ARGA

**Comment by rusty1s** [comment]
Hi, the paper you mentioned seems to be relatively similar to [VGAEs](https://arxiv.org/abs/1611.07308), where the variational autoencoder is a replaced with an adversarial autoencoder. This can be built in PyTorch Geometric within a few lines of code. I might make an example to showcase this. Other papers I am aware of are [GraphGANs](https://arxiv.org/abs/1711.08267) or [NetGANs](http://proceedings.mlr.press/v80/bojchevski18a/bojchevski18a.pdf), but I am sure there are tons of other ones.

**Comment by jlevy44** [comment]
@rusty1s @robclouth I second this. I'd really like an exploration of generative methods using torch_geometric.

**Comment by jlevy44** [comment]
I can help write the code if you are busy. If you have practical recommendations for writing it, I'd love to hear before this becomes a time sink.

**Comment by jlevy44** [comment]
I may just work on implementing ARGA if you haven't already.

**Comment by jlevy44** [comment]
I have a small network that I'd like to compare to ERGM and LatentNet models

**Comment by rusty1s** [comment]
Since request is growing, it's definitively time to integrate graph generation methods into PyG. VGAE + ARGA models/examples seem to be a good starting point before implementing more time-consuming ones. Maybe we should add a new folder `torch_geometric/models` to integrate those?

@jlevy44 Would love to see you making the first move. If you have any problems, feel free to reach out. Otherwise, I could implement it in the next week.

**Comment by jlevy44** [comment]
Sure. I'll give ARGA a shot and let you know if I run into issues. May take me more than a week though since I'm leaving for a trip tomorrow.

**Comment by jlevy44** [comment]
Do you have any recommendations for similar architectures that operate autoregressively? Maybe I can just define a loss function that takes into account random walks. Goal is to capture transitive properties of graph.

**Comment by jlevy44** [comment]
Maybe some combination between netGAN and ARGA would be nice to implement in the future.

**Comment by rusty1s** [comment]
[GraphRNN](https://arxiv.org/pdf/1802.08773.pdf)?

**Comment by jlevy44** [comment]
Yep. I agree with this. I've read the paper, been wanting to implement, how easy is it to implement?

**Comment by jlevy44** [comment]
I haven't tested my code yet, but does this seem like it's on the right track??:

```
def calc_loss(A_orig, A_new, d_real, d_fake, weights):# weights = (float(A_orig.shape[0]**2-A_orig.sum())/A_orig.sum()).flatten()
    dc_real_loss=nn.BCELoss(reduction='mean')(torch.ones(d_real.size()),d_real)
    dc_fake_loss=nn.BCELoss(reduction='mean')(torch.zeros(d_fake.size()),d_fake)
    A_loss = nn.BCELoss(reduction='sum',weight)(A_orig,A_new)
    total_loss = dc_real_loss + dc_fake_loss + A_loss
    return total_loss
    
def calc_weights(A):
    return (float(A.shape[0]**2-A.sum())/A.sum()).flatten()

def nx_graph2matrix(G):
    return nx.to_sparse_matrix(G).todense()

def X_df_to_tensor(x_df):
    return torch.tensor(x_df.values, dtype=torch.float)

def graph_to_data(graph_dict, y=None):
    A=torch.tensor(nx_graph2matrix(graph_dict['A']))
    index, value=dense_to_sparse(A)
    return Data(x=X_df_to_tensor(graph_dict['X']), edge_index=index, edge_attr=value, y=y)

def return_adjacency(geom_data):
    edge_index, edge_attr = geom_data.edge_index, geom_data.edge_attr
    return sparse_to_dense(edge_index, edge_attr)


class EncoderGCN(nn.Module):
    def __init__(self, n_total_features, n_latent, p_drop=0.):
        super(EncoderGCN, self).__init__()
        self.p_drop = p_drop
        self.n_total_features = n_total_features
        self.conv1 = nn.Sequential(torch_geometric.nn.GCNConv(self.n_total_features, 16),nn.ReLU(),nn.Dropout(self.p_drop))
        self.conv2 = nn.Sequential(torch_geometric.nn.GCNConv(16, 16),nn.ReLU(),nn.Dropout(self.p_drop))
        self.conv3 = torch_geometric.nn.GCNConv(16,n_latent)

    def forward(self, data):
        x, edge_index = data.x, data.edge_index
        x = self.conv1(x, edge_index) # X,A
        x = self.conv2(x, edge_index)
        x = self.conv3(x, edge_index)
        return x
    
class DecoderGCN(nn.Module):
    def __init__(self):
        super(DecoderGCN, self).__init__()
        self.output_layer = nn.Sigmoid()
        
    def forward(self, z):
        A=self.output_layer(torch.mm(z,torch.t(z)))
        return A
    
class Discriminator(nn.Module):
    def __init__(self, n_input):
        super(Distriminator, self).__init__()
        self.fc1 = nn.Sequential(nn.Linear(n_input,40), nn.ReLU())
        self.fc2 = nn.Sequential(nn.Linear(40,30), nn.ReLU())
        self.fc3 = nn.Sequential(nn.Linear(30,1),nn.Sigmoid())
        
    def forward(self, z):
        z=self.fc1(z)
        z=self.fc2(z)
        out=self.fc3(z)
        return out
    
class ARGA(nn.Module):
    def __init__(self, n_total_features, n_latent):
        self.encoder = EncoderGCN(n_total_features, n_latent)
        self.decoder = DecoderGCN()
        self.discriminator = Discriminator(n_latent)
        
    def forward(self, x):
        z_real=torch.randn(z_fake.size())
        z_fake = self.encoder(x)
        A = self.decoder(z_fake)
        d_real=self.discriminator(z_real)
        d_fake=self.discriminator(z_fake)
        return A, d_real, d_fake
    
    def simulate(self, x):
        z_fake = self.encoder(x)
        A = self.decoder(z_fake)
        return A
    
class ARGA_Model:
    def __int__(self, optimizer, n_epochs, model=ARGA, n_nodes=20, n_latent=30, weights=None, scheduler_opts=dict(scheduler='null',lr_scheduler_decay=0.5,T_max=10,eta_min=5e-8,T_mult=2)):
        self.loss_fn = lambda A_orig, A_new, d_real, d_fake: calc_loss(A_orig, A_new, d_real, d_fake, weights)
        self.model=model(n_nodes,n_latent)
        self.optimizer = optimizer
        self.scheduler = Scheduler(optimizer, scheduler_opts)
        self.n_epochs = n_epochs
        self.flatten = nn.Flatten()
        
    def train(self, data): # one
        self.model.train()
        total_loss = []
        for epoch in range(len(self.n_epochs)):
            A_real = return_adjacency(data)
            A_fake, d_real, d_fake = self.model(forward)
            A_real, A_fake = self.flatten(A_real), self.flatten(A_fake)
            loss = self.loss_fn(A_orig, A_new, d_real, d_fake)
            loss.backward()
            optimizer.step()
            scheduler.step()
            total_loss.append(loss.item())
            
    def simulate(self, data):
        self.model.eval()
        return return_adjacency(self.model.simulate(data)).item()
```

**Comment by jlevy44** [comment]
@rusty1s If you could just check superficially from the paper: https://arxiv.org/pdf/1802.04407.pdf

**Comment by jlevy44** [comment]
Not sure if I should employ a data loader if it is just for a single graph...

**Comment by jlevy44** [comment]
This is for the dataset to be loaded:

```
class SingleGraph(Dataset):
    def __init__(self, graph_dict, outcome_col=None): # if None do all vs all
        """graph_dict=dict(A=nx.Graph(...),X=pd.DataFrame(..., index=[node labels], columns=[covariates]))"""
        self.graph = graph_dict
        self.all_labels = np.array(list(G.nodes()))
        self.graph=graph_to_data(self.graph, outcome_col)
        self.length = len(self.all_labels)
        self.y = outcome_col
        if self.y==None:
            self.y = np.ones((self.length,))
        
    def __getitem__(self, i):
        return self.graph
        
    def __len__(self):
        return self.length
```

**Comment by jlevy44** [comment]
I'll try to wrap this up after you provide some feedback. If anything, we can make a PR and continue this discussion over there.

**Comment by jlevy44** [comment]
I'll try to test in the meanwhile and see if I can get anything to work.

**Comment by rusty1s** [comment]
You are fast :) The basic idea looks correct to me. However, it would be really cool to also provide mini-batch support (this should be possible with `torch_geometric.utils.to_dense_batch()` after encoding). I can give you in-depth feedback tomorrow.

**Comment by jlevy44** [comment]
Okay thanks. I just put this together in about an hour and a half so I don't expect any of this to be functional. Thanks! If you could give me a list of things to do to wrap it up, that'd be awesome. 

Hope we can PR it soon so I can start using it in my work.

**Comment by jlevy44** [comment]
Ah I see I made a few mistakes up at the sequential layers of the encoder (can't accept edge_list because I wrapped it into sequential). I'm sure you'll pick it all up. I'm currently debugging, working my way through this code. Looking forward to your feedback!

**Comment by jlevy44** [comment]
Actually got it to run the simulate method, so this could be promising.

**Comment by jlevy44** [comment]
I was able to train and test the model. Results didn't look so great though. Train loss dropped from 50k to 1k, but simulated graph appeared to be quite off (all of the values dropped to near 0.5). I think it's an implementation error, but would appreciate the direction.

**Comment by jlevy44** [comment]
I wonder if the adversarial regularization needs to be given higher weight, similar to VAE KL loss for example. Once this works, I'll add KL loss and shoot for VGAE.

**Comment by rusty1s** [comment]
Cool! Just wondering why you use `reduction='sum'` in the reconstruction loss. This is not done in the [official implementation](https://github.com/tkipf/gae/blob/master/gae/optimizer.py). Can you open a pull request and discuss the details there?

**Comment by jlevy44** [comment]
Sure, sounds good.

**Comment by jlevy44** [comment]
Shall we close this thread?
